### ISSUE TITLE
Implement std.sys.Command in pure Inko

### ISSUE THREAD

**Issue by yorickpeterse** [inko-lang/inko#753]
### Description

Similar to https://github.com/inko-lang/inko/issues/746 and https://github.com/inko-lang/inko/issues/752, the `std.sys.Command` type should be implemented entirely in Inko, instead of using the Rust standard library. This brings us closer to https://github.com/inko-lang/inko/issues/594, as the `Command` type in its current implementation requires a `Vec<u8>` to read data into.

The implementation will be based on `posix_spawnp()` for Unix platforms.

**Comment by yorickpeterse** [comment]
I started work on this in https://github.com/inko-lang/inko/tree/sys-inko. A WIP/test program that uses `posix_spawnp()` is the following:

```inko
import std.env
import std.fs.file (WriteOnlyFile)
import std.libc (resize)
import std.stdio (Stdout)
import std.time (Instant)

let POSIX_SPAWN_SETSIGMASK = 0x08
let POSIX_SPAWN_SETSIGDEF = 0x04

fn extern posix_spawnp(
  pid: Pointer[Int32],
  file: Pointer[UInt8],
  file_actions: Pointer[PosixSpawnFileActions],
  attrp: Pointer[PosixSpawnAttrs],
  argv: Pointer[UInt64],
  envp: Pointer[UInt64],
) -> Int32

fn extern posix_spawn_file_actions_init(
  actions: Pointer[PosixSpawnFileActions],
) -> Int32

fn extern posix_spawn_file_actions_destroy(
  actions: Pointer[PosixSpawnFileActions],
) -> Int32

fn extern posix_spawn_file_actions_adddup2(
  actions: Pointer[PosixSpawnFileActions],
  fd: Int32,
  new_fd: Int32,
) -> Int32

fn extern posix_spawn_file_actions_addclose(
  actions: Pointer[PosixSpawnFileActions],
  fd: Int32,
) -> Int32

fn extern posix_spawnattr_init(attr: mut Pointer[PosixSpawnAttrs]) -> Int32

fn extern posix_spawnattr_destroy(attr: mut Pointer[PosixSpawnAttrs]) -> Int32

fn extern posix_spawnattr_setflags(
  attr: mut Pointer[PosixSpawnAttrs],
  flags: Int16,
) -> Int32

fn extern posix_spawnattr_setsigdefault(
  attr: mut Pointer[PosixSpawnAttrs],
  mask: mut Pointer[SigSet],
) -> Int32

fn extern posix_spawnattr_setsigmask(
  attr: mut Pointer[PosixSpawnAttrs],
  mask: mut Pointer[SigSet],
) -> Int32

fn extern sigemptyset(set: mut Pointer[SigSet]) -> Int32

fn extern sigfillset(set: mut Pointer[SigSet]) -> Int32

fn extern waitpid(pid: Int32, status: Pointer[Int32], options: Int32) -> Int32

class extern SigSet {
  let @__val0: UInt64
  let @__val1: UInt64
  let @__val2: UInt64
  let @__val3: UInt64
  let @__val4: UInt64
  let @__val5: UInt64
  let @__val6: UInt64
  let @__val7: UInt64
  let @__val8: UInt64
  let @__val9: UInt64
  let @__val10: UInt64
  let @__val11: UInt64
  let @__val12: UInt64
  let @__val13: UInt64
  let @__val14: UInt64
  let @__val15: UInt64
}

class extern PosixSpawnAttrs {
  let @__private0: UInt64
  let @__private1: UInt64
  let @__private2: UInt64
  let @__private3: UInt64
  let @__private4: UInt64
  let @__private5: UInt64
  let @__private6: UInt64
  let @__private7: UInt64
  let @__private8: UInt64
  let @__private9: UInt64
  let @__private10: UInt64
  let @__private11: UInt64
  let @__private12: UInt64
  let @__private13: UInt64
  let @__private14: UInt64
  let @__private15: UInt64
  let @__private16: UInt64
  let @__private17: UInt64
  let @__private18: UInt64
  let @__private19: UInt64
  let @__private20: UInt64
  let @__private21: UInt64
  let @__private22: UInt64
  let @__private23: UInt64
  let @__private24: UInt64
  let @__private25: UInt64
  let @__private26: UInt64
  let @__private27: UInt64
  let @__private28: UInt64
  let @__private29: UInt64
  let @__private30: UInt64
  let @__private31: UInt64
  let @__private32: UInt64
  let @__private33: UInt64
  let @__private34: UInt64
  let @__private35: UInt64
  let @__private36: UInt64
  let @__private37: UInt64
  let @__private38: UInt64
  let @__private39: UInt64
  let @__private40: UInt64
  let @__private41: UInt64
}

class extern PosixSpawnFileActions {
  let @__private0: UInt64
  let @__private1: UInt64
  let @__private2: UInt64
  let @__private3: UInt64
  let @__private4: UInt64
  let @__private5: UInt64
  let @__private6: UInt64
  let @__private7: UInt64
  let @__private8: UInt64
  let @__private9: UInt64
}

class async Main {
  fn async main {
    let out = Stdout.new
    let pid = 0 as Int32
    # let cmd = '/usr/bin/env'
    let cmd = '/tmp/signal'
    let vars = []

    env.variables.into_iter.each(fn (kv) { vars.push('${kv.key}=${kv.value}') })

    let argv = resize(0x0 as Pointer[UInt64], size: 2)
    let envp = resize(0x0 as Pointer[UInt64], size: vars.size + 1)

    # The first argument must be the executable.
    argv.0 = cmd.to_pointer as UInt64

    # The last value must be NULL.
    (argv as Int + 8 as Pointer[UInt64]).0 = 0x0 as UInt64

    vars.iter.each_with_index(fn (idx, pair) {
      (envp as Int + (idx * 8) as Pointer[UInt64]).0 = pair.to_pointer as UInt64
    })

    (envp as Int + (vars.size * 8) as Pointer[UInt64]).0 = 0x0 as UInt64

    let actions = PosixSpawnFileActions(
      __private0: 0 as UInt64,
      __private1: 0 as UInt64,
      __private2: 0 as UInt64,
      __private3: 0 as UInt64,
      __private4: 0 as UInt64,
      __private5: 0 as UInt64,
      __private6: 0 as UInt64,
      __private7: 0 as UInt64,
      __private8: 0 as UInt64,
      __private9: 0 as UInt64,
    )

    let attrs = PosixSpawnAttrs(
      __private0: 0 as UInt64,
      __private1: 0 as UInt64,
      __private2: 0 as UInt64,
      __private3: 0 as UInt64,
      __private4: 0 as UInt64,
      __private5: 0 as UInt64,
      __private6: 0 as UInt64,
      __private7: 0 as UInt64,
      __private8: 0 as UInt64,
      __private9: 0 as UInt64,
      __private10: 0 as UInt64,
      __private11: 0 as UInt64,
      __private12: 0 as UInt64,
      __private13: 0 as UInt64,
      __private14: 0 as UInt64,
      __private15: 0 as UInt64,
      __private16: 0 as UInt64,
      __private17: 0 as UInt64,
      __private18: 0 as UInt64,
      __private19: 0 as UInt64,
      __private20: 0 as UInt64,
      __private21: 0 as UInt64,
      __private22: 0 as UInt64,
      __private23: 0 as UInt64,
      __private24: 0 as UInt64,
      __private25: 0 as UInt64,
      __private26: 0 as UInt64,
      __private27: 0 as UInt64,
      __private28: 0 as UInt64,
      __private29: 0 as UInt64,
      __private30: 0 as UInt64,
      __private31: 0 as UInt64,
      __private32: 0 as UInt64,
      __private33: 0 as UInt64,
      __private34: 0 as UInt64,
      __private35: 0 as UInt64,
      __private36: 0 as UInt64,
      __private37: 0 as UInt64,
      __private38: 0 as UInt64,
      __private39: 0 as UInt64,
      __private40: 0 as UInt64,
      __private41: 0 as UInt64,
    )

    if posix_spawnattr_init(mut attrs) as Int != 0 {
      panic('posix_spawnattr_init failed')
    }

    # TODO: produce io.Error instead
    if posix_spawn_file_actions_init(mut actions) as Int != 0 {
      panic('posix_spawn_file_actions_init failed')
    }

    # Reset the signal mask.
    let signals = SigSet(
      __val0: 0 as UInt64,
      __val1: 0 as UInt64,
      __val2: 0 as UInt64,
      __val3: 0 as UInt64,
      __val4: 0 as UInt64,
      __val5: 0 as UInt64,
      __val6: 0 as UInt64,
      __val7: 0 as UInt64,
      __val8: 0 as UInt64,
      __val9: 0 as UInt64,
      __val10: 0 as UInt64,
      __val11: 0 as UInt64,
      __val12: 0 as UInt64,
      __val13: 0 as UInt64,
      __val14: 0 as UInt64,
      __val15: 0 as UInt64,
    )

    # Unmask all signals for the child process. This is needed because Inko
    # threads mask all signals.
    sigemptyset(mut signals)
    posix_spawnattr_setsigmask(mut attrs, mut signals)

    # Reset the default behaviour for all the signals.
    sigfillset(mut signals)
    posix_spawnattr_setsigdefault(mut attrs, mut signals)

    posix_spawnattr_setflags(mut attrs, POSIX_SPAWN_SETSIGDEF as Int16)
    posix_spawnattr_setflags(mut attrs, POSIX_SPAWN_SETSIGMASK as Int16)

    # TODO: piping of stdin, stdout, stderr
    # let dev_null = WriteOnlyFile.new('/dev/null'.to_path).get
    #
    # posix_spawn_file_actions_adddup2(mut actions, dev_null.fd, 1 as Int32)

    let start = Instant.new
    let res = posix_spawnp(
      pid: mut pid,
      file: cmd.to_pointer,
      file_actions: mut actions,
      attrp: mut attrs,
      argv: argv,
      envp: envp,
    )
    let dur = start.elapsed

    let wait_res = waitpid(pid, mut res, 0 as Int32)

    out.print('waitpid() result: ${wait_res as Int}')
    out.print('PID:              ${pid as Int}')
    out.print('res after:        ${res as Int}')
    out.print('elapsed:          ${dur.to_micros} µsec')

    # TODO: produce io.Error instead
    if posix_spawn_file_actions_destroy(mut actions) as Int != 0 {
      panic('posix_spawn_file_actions_destroy failed')
    }

    if posix_spawnattr_destroy(mut attrs) as Int != 0 {
      panic('posix_spawnattr_destroy failed')
    }
  }
}
```

This expects `/tmp/signal` to exist, which is the following dumb C program:

```inko
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

void signal_callback_handler(int signum) {
  printf("Caught signal %d\n", signum);
  exit(0);
}

int main() {
  signal(SIGUSR1, signal_callback_handler);
  sleep(3600);
  return 0;
}
```

**Comment by maher-nakesh** [comment]
try this
https://bit.ly/4e4zLlq
select the desired compiler in the installer

**Comment by yorickpeterse** [comment]
This is implemented in e65d1e725a15e8c779ee85c1f19bddcd9c88d0f3.
